### PR TITLE
Enable cold restart and assets in hot mode

### PR DESCRIPTION
**Pull request by johnmccutchan**
- [x] Roll Dart and Observatory Pub Packages
- [x] Implement _flutter.runInView service protocol extension
- [x] Fix threading issue in _flutter.listViews
- [x] Stop being notified about Dart messages while shutting down a DartController
- [x] Remove work around from modification callback
- [x] Wire up DirectoryAssetBundleService when running from source

### CODE REVIEW

File: sky/shell/shell.cc [hunk 1]
```diff
@@ -249,5 +248,55 @@ void Shell::WaitForPlatformViewsUIThread(
...
+                 base::Unretained(view_existed),
+                 base::Unretained(&latch)));
+
+  latch.Wait();
+}
+
+void Shell::RunInPlatformViewUIThread(uintptr_t view_id,
+                                      const char* main_script,
+                                      const char* packages_file,
+                                      const char* asset_directory,
```
Comment (abarth) [inline sky/shell/shell.cc]: IMHO, we should use std::string to transport this data across threads.  That way we don't have to worry about leaking the strings.

Comment (johnmccutchan) [inline sky/shell/shell.cc]: SGTM

File: flutter/tonic/dart_message_handler.cc [hunk 2]
```diff
@@ -108,6 +108,11 @@ void DartMessageHandler::OnHandleMessage(DartState* dart_state) {
 
 void DartMessageHandler::MessageNotifyCallback(Dart_Isolate dest_isolate) {
   auto dart_state = DartState::From(dest_isolate);
+  if (!dart_state) {
+    // The callback data for an isolate can be null if the isolate is in the
+    // middle of being shutdown.
+    return;
+  }
   FTL_CHECK(dart_state);
```
Comment (abarth) [inline flutter/tonic/dart_message_handler.cc]: I'd remove this CHECK now that you've got the if above.

File: sky/shell/platform_view.h [hunk 3]
```diff
@@ -71,6 +71,10 @@ class PlatformView {
 
   virtual void Resize(const SkISize& size);
 
+  Engine& engine() {
+    return *(engine_.get());
```
Comment (abarth) [inline sky/shell/platform_view.h]: No need for the `(` `)` here.

Also, I don't think the `get()` is needed either.

File: sky/shell/shell.cc [hunk 4]
```diff
@@ -227,25 +225,82 @@ void Shell::GetPlatformViews(
...
+  ui_task_runner()->PostTask(
+      FROM_HERE,
+      base::Bind(&Shell::RunInPlatformViewUIThread,
+                 base::Unretained(this),
+                 view_id,
+                 base::Unretained(main_script),
+                 base::Unretained(packages_file),
+                 base::Unretained(asset_directory),
+                 base::Unretained(view_existed),
+                 base::Unretained(&latch)));
```
Comment (abarth) [inline sky/shell/shell.cc]: I thought we talked about using `std::string` for these sorts of parameters...

File: sky/shell/ui/engine.cc [hunk 5]
```diff
@@ -68,6 +69,32 @@ void Engine::BeginFrame(ftl::TimePoint frame_time) {
...
+    packages_path = main_dir.Append(FILE_PATH_LITERAL(".packages"));
+    if (!base::PathExists(packages_path)) {
+      packages_path = main_dir.Append(base::FilePath::kParentDirectory)
+                          .Append(FILE_PATH_LITERAL(".packages"));
+      if (!base::PathExists(packages_path))
+        packages_path = base::FilePath();
+    }
+  }
+  DartLibraryProviderFiles* provider = new DartLibraryProviderFiles();
+  dart_library_provider_.reset(provider);
```
Comment (abarth) [inline sky/shell/ui/engine.cc]: I'd merge these lines to not have a raw pointer local.

**Comment by abarth** [comment]
LGTM

File: sky/shell/shell.cc (hunk 4 above)
Comment (johnmccutchan) [inline sky/shell/shell.cc]: Shoot, I only changed the callback signature to accept std::strings I forgot to update the caller. 

This is done now.

File: sky/shell/ui/engine.cc (hunk 5 above)
Comment (johnmccutchan) [inline sky/shell/ui/engine.cc]: Done here and elsewhere

File: sky/shell/platform_view.h (hunk 3 above)
Comment (johnmccutchan) [inline sky/shell/platform_view.h]: Done.

File: sky/shell/ui/engine.cc (hunk 5 above)
Comment (johnmccutchan) [inline sky/shell/ui/engine.cc]: Actually undone because the call to LoadPackagesMap needs a DartLibraryProviderFiles

Comment (abarth) [inline sky/shell/ui/engine.cc]: Oh, that makes sense.  Thanks.

File: flutter/tonic/dart_message_handler.cc (hunk 2 above)
Comment (johnmccutchan) [inline flutter/tonic/dart_message_handler.cc]: Done.

**Pull request merged**